ze/btx_zeinterval_callbacks.cpp

```diff
@@ -0,0 +1,727 @@
+#include "xprof_utils.hpp"
+#include <btx_zeinterval_callbacks.hpp>
+#include <cassert>
+#include <iostream>
+#include <metababel/metababel.h>
+#include <regex>
+#include <stdio.h>
+#include <string>
+#include <unordered_map>
+
+std::string strip_event_class_name(const char *str) {
+  std::string temp(str);
+  std::smatch match;
+  std::regex_search(temp, match, std::regex(":(.*?)_?(?:entry|exit)?$"));
+  assert(match.size() > 1 && "Event_class_name not matching regex.");
+  return match[1].str();
+}
+
+/*
+ * Memory Interval
+ */
+
+static void add_memory(data_t *state, hp_t hp, uintptr_t ptr, size_t size,
+                       std::string source) {
+  std::unordered_map<hp_t, memory_interval_t> *mi = nullptr;
+  if (source == "lttng_ust_ze:zeMemAllocHost_exit")
+    mi = &state->rangeset_memory_host;
+  else if (source == "lttng_ust_ze:zeMemAllocDevice_exit")
+    mi = &state->rangeset_memory_device;
+  else if (source == "lttng_ust_ze:zeMemAllocShared_exit")
+    mi = &state->rangeset_memory_shared;
+  else
+    std::cout << "WARNING Adding unknow memory " << source << "ptr " << ptr
+              << std::endl;
+
+  (*mi)[hp][ptr] = ptr + size;
+}
```

Comment:
```suggestion
static void add_memory(data_t *state, hp_t hp, uintptr_t ptr, size_t size,
                       std::string source) {
  switch(source) {
    case "lttng_ust_ze:zeMemAllocHost_exit":
      state->rangeset_memory_host[hp][ptr] = ptr + size;
      break;
    case "lttng_ust_ze:zeMemAllocDevice_exit":
      state->rangeset_memory_device[hp][ptr] = ptr + size;
      break;
    case "lttng_ust_ze:zeMemAllocShared_exit"":
      state->rangeset_memory_shared[hp][ptr] = ptr + size;
      break;
    default:
      std::cout << "WARNING Adding unknown memory " << source << "ptr " << ptr << std::endl;
  }
}
```

Reply:
I think (not 100% sure) in C++ the switch should be able to manage string comparison. Also, I did not understand why you needed a temporary variable `m`, since you can update the values directly, perhaps I am wrong.

Reply:
Don't wnt to duplicate `  (*mi)[hp][ptr] = ptr + size;`